testsuite/cpptests/test_sort.h

```diff
@@ -27,78 +27,151 @@
 #include <boost/test/unit_test.hpp>
 
 // C++ includes:
+#include <algorithm>
 #include <vector>
 
 // Includes from libnestutil:
 #include "sort.h"
 
 namespace nest
 {
-
+/**
+ * Wrapper for quicksort3way.
+ *
+ * When calling nest::sort() directly it is impossible to sort with the
+ * built in quicksort3way from tests. This is because if NEST is compiled
+ * with Boost support, nest::sort() sorts with Boost, and if NEST is not
+ * compiled with Boost, nest::sort() calls quicksort3way, but the Boost
+ * tests are not compiled.
+ */
 void
-nest_quicksort( BlockVector< size_t >& bv0, BlockVector< size_t >& bv1 )
+nest_quicksort( BlockVector< int >& bv0, BlockVector< int >& bv1 )
 {
   nest::quicksort3way( bv0, bv1, 0, bv0.size() - 1 );
 }
 
-bool
-is_sorted( BlockVector< size_t >::const_iterator begin, BlockVector< size_t >::const_iterator end )
+/**
+ * Fixture filling two BlockVectors and a vector with lineary decreasing numbers.
+ * The vector is then sorted.
+ */
+struct fill_bv_vec_linear
 {
-  for ( BlockVector< size_t >::const_iterator it = begin; it < --end; )
+  fill_bv_vec_linear()
+    : N( 20000 )
+    , bv_sort( N )
+    , bv_perm( N )
+    , vec_sort( N )
   {
-    if ( *it > *( ++it ) )
+    for ( int i = 0; i < N; ++i )
     {
-      return false;
+      int element = N - i;
+      bv_sort[ i ] = element;
+      bv_perm[ i ] = element;
+      vec_sort[ i ] = element;
     }
+    std::sort( vec_sort.begin(), vec_sort.end() );
+  }
+
+  const int N;
+  BlockVector< int > bv_sort;
+  BlockVector< int > bv_perm;
+  std::vector< int > vec_sort;
+};
+
+/**
+ * Fixture filling two BlockVectors and a vector with random numbers.
+ * The vector is then sorted.
+ */
+struct fill_bv_vec_random
+{
+  fill_bv_vec_random()
+    : N( 20000 )
+    , bv_sort( N )
+    , bv_perm( N )
+    , vec_sort( N )
+  {
+    for ( int i = 0; i < N; ++i )
+    {
+      const size_t k = std::rand() % N;
+      bv_sort[ i ] = k;
+      bv_perm[ i ] = k;
+      vec_sort[ i ] = k;
+    }
+    std::sort( vec_sort.begin(), vec_sort.end() );
   }
-  return true;
-}
+
+  const int N;
+  BlockVector< int > bv_sort;
+  BlockVector< int > bv_perm;
+  std::vector< int > vec_sort;
+};
 
 BOOST_AUTO_TEST_SUITE( test_sort )
 
 /**
  * Tests whether two arrays with randomly generated numbers are sorted
- * correctly by a single call to sort.
+ * correctly when sorting with the built-in quicksort.
  */
-BOOST_AUTO_TEST_CASE( test_random )
+BOOST_FIXTURE_TEST_CASE( test_quicksort_random, fill_bv_vec_random )
 {
-  const size_t N = 20000;
-  BlockVector< size_t > bv0( N );
-  BlockVector< size_t > bv1( N );
+  nest_quicksort( bv_sort, bv_perm );
 
-  for ( size_t i = 0; i < N; ++i )
-  {
-    const size_t k = std::rand() % N;
-    bv0[ i ] = k;
-    bv1[ i ] = k;
-  }
+  BOOST_REQUIRE( std::is_sorted( bv_sort.begin(), bv_sort.end() ) );
+  BOOST_REQUIRE( std::is_sorted( bv_perm.begin(), bv_perm.end() ) );
+
```

Comment:
Could you add a comment explaining why you perform both the `is_sorted` and the `equal` tests?